pkg/kubectl/cmd/auth/reconcile.go

```diff
@@ -0,0 +1,207 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"errors"
+	"io"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/apis/rbac"
+	internalcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	internalrbacclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/rbac/internalversion"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/registry/rbac/reconciliation"
+)
+
+// ReconcileOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type ReconcileOptions struct {
+	ResourceBuilder *resource.Builder
+	RBACClient      internalrbacclient.RbacInterface
+	CoreClient      internalcoreclient.CoreInterface
+
+	Print func(*resource.Info) error
+
+	Out io.Writer
+	Err io.Writer
+}
+
+var (
+	reconcileLong = templates.LongDesc(`
+		Reconciles rules for RBAC Role, RoleBinding, ClusterRole, and ClusterRole binding objects`)
+
+	reconcileExample = templates.Examples(`
+		# Reconcile rbac resources from a file
+		kubectl auth reconcile -f my-rbac-rules.yaml`)
+)
+
+func NewCmdReconcile(f cmdutil.Factory, out, err io.Writer) *cobra.Command {
+	fileOptions := &resource.FilenameOptions{}
+	o := &ReconcileOptions{
+		Out: out,
+		Err: err,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "reconcile -f [FILE|URL|DIRECTORY]",
+		Short:   "Reconciles rules for RBAC Role, RoleBinding, ClusterRole, and ClusterRole binding objects",
+		Long:    reconcileLong,
+		Example: reconcileExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete(cmd, f, args, fileOptions))
+			cmdutil.CheckErr(o.Validate())
+			cmdutil.CheckErr(o.RunReconcile())
+		},
+	}
+
+	cmdutil.AddPrinterFlags(cmd)
+	usage := "identifying the resource to reconcile."
+	cmdutil.AddFilenameOptionFlags(cmd, fileOptions, usage)
+
+	return cmd
+}
+
+func (o *ReconcileOptions) Complete(cmd *cobra.Command, f cmdutil.Factory, args []string, options *resource.FilenameOptions) error {
+	if len(args) > 0 {
+		return errors.New("no arguments are allowed")
+	}
+
+	namespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	o.ResourceBuilder = f.NewBuilder(true).
+		ContinueOnError().
+		NamespaceParam(namespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, options).
+		ResourceTypeOrNameArgs(false, args...).
```

Comment:
`ResourceTypeOrNameArgs` lets them specify a resource name that is fetched from the server? what would that do for reconcile?

Reply:
> ResourceTypeOrNameArgs lets them specify a resource name that is fetched from the server? what would that do for reconcile?

nothing since args are forced to be zero length.  I'll remove that bit though.